Mage.Sets/src/mage/cards/j/JaredCarthalion.java

```diff
@@ -0,0 +1,135 @@
+
+        // −3: Choose up to two target creatures. For each of them, put a number of +1/+1 counters on it equal to the number of colors it is.
+        Ability ability1 = new LoyaltyAbility(new JaredCarthalionBoostEffect(), -3);
+        ability1.addTarget(new TargetCreaturePermanent(0, 2));
+        this.addAbility(ability1);
+
+        // −6: Return target multicolored card from your graveyard to your hand. If that card was all colors, draw a card and create two Treasure tokens.
+        Ability ability2 = new LoyaltyAbility(new JaredCarthalionUltimateEffect(), -6);
+        ability2.addTarget(new TargetCardInYourGraveyard(filter));
+        this.addAbility(ability2);
+
+        // Jared Carthalion can be your commander.
+        this.addAbility(CanBeYourCommanderAbility.getInstance());
+    }
+
+    private JaredCarthalion(final JaredCarthalion card) {
+        super(card);
+    }
+
+    @Override
+    public JaredCarthalion copy() {
+        return new JaredCarthalion(this);
+    }
+}
+
+class JaredCarthalionBoostEffect extends OneShotEffect {
+
+    public JaredCarthalionBoostEffect() {
+        super(Outcome.BoostCreature);
+        this.staticText = "Choose up to two target creatures. For each of them, put " +
+                "a number of +1/+1 counters on it equal to the number of colors it is.";
+    }
+
+    public JaredCarthalionBoostEffect(final JaredCarthalionBoostEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public JaredCarthalionBoostEffect copy() {
+        return new JaredCarthalionBoostEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        for (UUID targetId : getTargetPointer().getTargets(game, source)) {
+            Permanent creature = game.getPermanent(targetId);
+            if (creature != null) {
+                Counter counter;
+                counter = CounterType.P1P1.createInstance(creature.getColor().getColorCount());
+                creature.addCounters(counter, source, game);
+            }
+        }
+        return true;
+    }
+}
+
+class JaredCarthalionUltimateEffect extends OneShotEffect {
+
+    public JaredCarthalionUltimateEffect() {
+        super(Outcome.Benefit);
+        this.staticText = "Return target multicolored card from your graveyard to your hand. " +
+                "If that card was all colors, draw a card and create two Treasure tokens.";
+    }
+
+    public JaredCarthalionUltimateEffect(final JaredCarthalionUltimateEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public JaredCarthalionUltimateEffect copy() {
+        return new JaredCarthalionUltimateEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        Player controller = game.getPlayer(source.getControllerId());
+        if (controller != null) {
+            Card card = game.getCard(targetPointer.getFirst(game, source));
+            if (card != null) {
```

Comment:
you should combine the card and player check and invert them to flatten the code here